core/storage/sqlite3_ondisk.rs

```diff
@@ -1650,205 +1653,306 @@ pub fn read_entire_wal_dumb(file: &Arc<dyn File>) -> Result<Arc<RwLock<WalFileSh
+            let h = self.header.lock();
+            (*h, st.use_native_endian)
         };
 
-        let mut cumulative_checksum = (header_locked.checksum_1, header_locked.checksum_2);
-        let page_size_u32 = header_locked.page_size;
-
-        if PageSize::new(page_size_u32).is_none() {
-            panic!("Invalid page size in WAL header: {page_size_u32}");
-        }
-        let page_size = page_size_u32 as usize;
-
-        let mut current_offset = WAL_HEADER_SIZE;
-        let mut frame_idx = 1_u64;
-
-        let mut wfs_data = wal_file_shared_for_completion.write();
-
-        if !checksum_header_failed {
-            while current_offset + WAL_FRAME_HEADER_SIZE + page_size <= buf_slice.len() {
-                let frame_header_slice =
-                    &buf_slice[current_offset..current_offset + WAL_FRAME_HEADER_SIZE];
-                let page_data_slice = &buf_slice[current_offset + WAL_FRAME_HEADER_SIZE
-                    ..current_offset + WAL_FRAME_HEADER_SIZE + page_size];
-
-                let frame_h_page_number =
-                    u32::from_be_bytes(frame_header_slice[0..4].try_into().unwrap());
-                let frame_h_db_size =
-                    u32::from_be_bytes(frame_header_slice[4..8].try_into().unwrap());
-                let frame_h_salt_1 =
-                    u32::from_be_bytes(frame_header_slice[8..12].try_into().unwrap());
-                let frame_h_salt_2 =
-                    u32::from_be_bytes(frame_header_slice[12..16].try_into().unwrap());
-                let frame_h_checksum_1 =
-                    u32::from_be_bytes(frame_header_slice[16..20].try_into().unwrap());
-                let frame_h_checksum_2 =
-                    u32::from_be_bytes(frame_header_slice[20..24].try_into().unwrap());
-
-                if frame_h_page_number == 0 {
-                    tracing::trace!(
-                        "WAL frame with page number 0. Ignoring frames starting from frame {}",
-                        frame_idx
-                    );
-                    break;
-                }
-                // It contains more frames with mismatched SALT values, which means they're leftovers from previous checkpoints
-                if frame_h_salt_1 != header_locked.salt_1 || frame_h_salt_2 != header_locked.salt_2
-                {
-                    tracing::trace!(
-                        "WAL frame salt mismatch: expected ({}, {}), got ({}, {}). Ignoring frames starting from frame {}",
-                        header_locked.salt_1,
-                        header_locked.salt_2,
-                        frame_h_salt_1,
-                        frame_h_salt_2,
-                        frame_idx
-                    );
-                    break;
-                }
+        let consumed = self.process_frames(buf_slice, &header_copy, use_native);
+        self.off_atomic.fetch_add(consumed as u64, Ordering::AcqRel);
+        // If we didn’t consume the full chunk, we hit a stop condition
```

Comment:
Can it be that the reason for early stop of process_frames were partial read which was not aligned to the wal frame size boundaries?